lib/src/modules/task.cc

```diff
@@ -426,9 +438,110 @@ static fs::path getCachedTaskFile(const fs::path& task_cache_dir,
     }
 }
 
+// build a spool dir for the supporting library files requested by a multifile task
+static fs::path createInstallDir(const fs::path& spool_dir, const std::set<std::string>& download_set) {
+    auto install_dir = spool_dir / fs::unique_path("temp_task_%%%%-%%%%-%%%%-%%%%");
+    fs::create_directories(install_dir);
+    fs::last_write_time(install_dir, time(nullptr));
+    fs::permissions(install_dir, NIX_DIR_PERMS);
+
+    // this should generate a unique collection of directory paths to append to install_dir
+    // for example:
+    // if files had the paths: /one/two/three.txt, /one/two/one.txt, /one/two.txt
+    // dir_structure set would contain: /one/two /one
+    std::set<fs::path> dir_structure;
+    for (auto f_name : download_set) {
+        dir_structure.insert(fs::path(f_name).parent_path());
+    }
+
+    // now create the directories to copy the required files into
+    for (auto dir_path : dir_structure) {
+        fs::path tmp = install_dir;
+        for (const fs::path &dir : dir_path) {
+            tmp /= dir;
+            fs::create_directories(tmp);
+            fs::last_write_time(tmp, time(nullptr));
+            fs::permissions(tmp, NIX_DIR_PERMS);
+        }
+    }
+
+    return install_dir;
+}
+
+// Get a unique list of "lib" files that the task has specified in metadata
+// this involves enumerating files when directories are requested
+std::set<std::string> getMultiFiles(std::vector<std::string> const& meta_files, std::vector<std::string>  const& impl_files, std::vector<lth_jc::JsonContainer> const& files) {
+    // build set of unique files to download and exclude the selected task executable
+    std::set<std::string> download_set(meta_files.begin(), meta_files.end());
+    download_set.insert(impl_files.begin(), impl_files.end());
+    std::vector<std::string> directories;
+
+    // expand files from directories specified by trailing '/'
+    for (auto f_name : download_set) {
```

Comment:
I'm not sure what you'll end up iterating over when you're also inserting items. This may do more work than needed, which is why I suggested the first pass just add to `directories`, then do a pass over directories to add items to `download_set`.

Reply:
Ahh, Yes, I see now (originally got rid of the `erase` calls but neglected to remove the `insert`) I understand that modifying the # of elements DS you are iterating over is not good... Updated.